src/Actions/Bulk/DeleteThreads.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace TeamTeaTime\Forum\Actions\Bulk;
+
+use Illuminate\Support\Facades\DB;
+use TeamTeaTime\Forum\Actions\BaseAction;
+use TeamTeaTime\Forum\Models\Thread;
+
+class DeleteThreads extends BaseAction
+{
+    private array $threadIds;
+    private bool $includeTrashed;
+    private bool $permaDelete;
+
+    public function __construct(array $threadIds, bool $includeTrashed, bool $permaDelete = false)
+    {
+        $this->threadIds = $threadIds;
+        $this->includeTrashed = $includeTrashed;
+        $this->permaDelete = $permaDelete;
+    }
+
+    protected function transact()
+    {
+        $query = Thread::whereIn('id', $this->threadIds);
+        $threads;
+
+        if ($this->includeTrashed)
+        {
+            $threads = $query->withTrashed()->get();
+
+            // Return early if this is a soft-delete and the selected threads are already trashed,
+            // or there are no valid threads in the selection
+            if (! $this->permaDelete && $threads->whereNull(Thread::DELETED_AT)->count() == 0) return null;
+        }
+        else
+        {
+            $threads = $query->get();
+
+            // Return early if there are no valid threads in the selection
+            if ($threads->count() == 0) return null;
+        }
+
+        // Use the raw query builder to prevent touching updated_at
+        $query = DB::table(Thread::getTableName())->whereIn('id', $this->threadIds);
+        $rowsAffected = $this->permaDelete
+            ? $query->delete()
+            : $query->whereNull(Thread::DELETED_AT)->update([Thread::DELETED_AT => DB::raw('now()')]);
+
+        $threadsByCategory = $threads->groupBy('category_id');
+        foreach ($threadsByCategory as $threads)
```

Comment:
This is confusing imo - you have a higher scoped variable called `$threads`, so there could easily be confusion over which is being used... I'd call it something like `$categoryThreads`.